numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -2079,18 +2080,24 @@ array_fromfile(PyObject *NPY_UNUSED(ignored), PyObject *args, PyObject *keywds)
     }
     ret = PyArray_FromFile(fp, type, (npy_intp) nin, sep);
 
+    /* If an exception is thrown in the call to PyArray_FromFile
+     * we need to clear it, and restore it later to ensure that
+     * we can cleanup the duplicated file descriptor properly.
+     */
+    PyErr_Fetch(&err_ptype, &err_pvalue, &err_ptraceback);
     if (npy_PyFile_DupClose2(file, fp, orig_pos) < 0) {
         goto fail;
     }
     if (own && npy_PyFile_CloseFile(file) < 0) {
         goto fail;
```

Comment:
Might be neat to call `_PyErr_ChainExceptions` here, although you'll need to add a backport of that in `numpy\core\include\numpy\npy_3kcompat.h` (almost identical to `npy_PyErr_ChainExceptionsCause`)

Reply:
This is part of the C-API that I'm not really familiar with - `npy_PyErr_ChainExceptionsCause` looks like it's almost identical to `_PyErr_ChainExceptions`.

What are the differences between them and why would I want to use one over the other?

Reply:
`_PyErr_ChainExceptions` isn't available in python 2.7, which is why you can't use it directly.

`_PyErr_ChainExceptions` gives the behavior of
```python
>>> try:
	raise ValueError
except:
	raise KeyError

Traceback (most recent call last):
  File "<pyshell#5>", line 2, in <module>
    raise ValueError
ValueError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<pyshell#5>", line 4, in <module>
    raise KeyError
KeyError
```

While `npy_PyErr_ChainExceptionsCause` gives the behavior of:
```python
>>> try:
	raise ValueError
except Exception as e:
	raise KeyError from e

Traceback (most recent call last):
  File "<pyshell#7>", line 2, in <module>
    raise ValueError
ValueError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<pyshell#7>", line 4, in <module>
    raise KeyError from e
KeyError
```

Reply:
Thanks for the explanation, I've added the implementation of that and have used it.